### ISSUE TITLE
Issue with Custom OPA Storage implementation

### ISSUE THREAD

**Issue by yadunandan004** [open-policy-agent/opa#6474]
[open-policy-agent/opa/rego/rego.go](https://github.com/open-policy-agent/opa/blob/main/rego/rego.go#L1778)
The above check here, prevents from implementing custom interface described here: https://github.com/open-policy-agent/opa/blob/main/storage/interface.go#L20

## Short description 
So when a new Rego instance is created (via https://github.com/open-policy-agent/opa/blob/main/rego/rego.go#L1197). A user is expected to provide modules during instantiation, even though the store interface implemented would already have the policies in it, rendering the modules param to be optional to the functionality. However, the provided modules parameter is still checked, and does not create a new instance.

Examples:

* OPA version : 0.59.0
``` 
func (a *Agent) newPolicyEngine(ctx context.Context) (*rego.Rego, error) {
	logger.Infof("creating new policy engine with latest policy and data")
	policyModules := make([]func(r *rego.Rego), 0)
	policyModules = append(policyModules, rego.Query(a.evalPath))
	policyModules = append(policyModules, rego.Store(a.myCustomStore)) // this already contains policy and data pulled from a server
	policyModules = append(policyModules, rego.Module("hello.rego", getHelloRego())) // dummy module added to sucessfully create a rego object
	return rego.New(policyModules...), nil
}

func getHelloRego() string {
	return `package play

import future.keywords.if

default hello := false

hello if input.message == "world"`
}
```
The above way of adding a dummy file (hello.rego) is the only way to circumvent the issue. It fails if this module is removed, even though this is unimportant to our business logic
-->

## Steps To Reproduce
* implement a store interface of your own
* add policy and data to it
* create the new rego object like above without the dummy rego file (hello.rego)

## Expected behavior
* opa should successfully create a new object with just a store interface, without having to check for modules.

## Additional context
* check if the store interface is also nil along with modules, and return error only if both are absent.

**Comment by ashutosh-narkar** [comment]
When you use the rego package, the assumption is that the compiler and store are kept in-sync. The store modules are only parsed if there are modules set on the rego object. More details about this can be found in [this](https://github.com/open-policy-agent/opa/commit/20612bb81943118fae581b06fc268f8207bdc6ec) commit.

**Comment by yadunandan004** [comment]
How would i ensure that compiler and store are in sync? I see that it is ensured for default store type (inmem), but not sure how to do this for a custom store implementation.  Could you please point to an example?

**Comment by ashutosh-narkar** [comment]
One way is when you add modules to the store, you'd compile them as well.

**Comment by yadunandan004** [comment]
okay but the line here : https://github.com/open-policy-agent/opa/blob/main/rego/rego.go#L1778 is still of type `rawModule`. Even if i insert compiled modules, it doesn't seem to change anything for me. So based on the code, it seems that i still need to pass a dummy module like this
`rego.Module("hello.rego", getHelloRego())` 
only after that, the modules in the compiler or store seem to be read (https://github.com/open-policy-agent/opa/blob/main/rego/rego.go#L1797).

**Comment by ashutosh-narkar** [comment]
Maybe I'm missing something, but if you have policies in the store that have been parsed and compiled, then the `parseModules` step would not be necessary as you've already done the steps that function does.
